File: packages/commons/octobot_commons/number_util.py
#  Drakkar-Software OctoBot-Commons
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import decimal
import math
import typing


def round_into_str_with_max_digits(number: float, digits_count: int) -> str:
    """
    Round the number with digits_count
    :param number: the number to round
    :param digits_count: the digit count
    :return: the rounded number
    """
    return "{:.{}f}".format(round(number, digits_count), digits_count)


def round_into_float_with_max_digits(number: float, digits_count: int) -> float:
    """
    Round the float number with digits_count
    :param number: the number to round
    :param digits_count: the digit count
    :return: the rounded number
    """
    return float(
        round_into_str_with_max_digits(number=number, digits_count=digits_count)
    )


def get_digits_count(value: typing.Union[float, decimal.Decimal]):
    """
    :return: the number of digits in the given number
    """
    return round(abs(math.log(value, 10)))

File: crates/octobot_commons/src/number_util.rs
pub fn round_into_str_with_max_digits(number: f64, digits_count: usize) -> String {
    let rounded = (number * 10f64.powi(digits_count as i32)).round() / 10f64.powi(digits_count as i32);
    format!("{:.prec$}", rounded, prec = digits_count)
}

pub fn round_into_float_with_max_digits(number: f64, digits_count: usize) -> f64 {
    round_into_str_with_max_digits(number, digits_count)
        .parse()
        .unwrap_or(number)
}

pub fn get_digits_count(value: f64) -> i64 {
    value.log10().abs().round() as i64
}
